lightning/src/ln/channel.rs

```diff
@@ -1719,9 +1795,50 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				removed_outbound_total_msat += htlc.amount_msat;
 			}
 		}
-		if htlc_inbound_value_msat + msg.amount_msat + self.value_to_self_msat > (self.channel_value_satoshis - Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis)) * 1000 + removed_outbound_total_msat {
-			return Err(ChannelError::Close("Remote HTLC add would put them under their reserve value"));
+
+		if self.channel_outbound {
+			// Check that they won't violate our local required channel reserve by adding this HTLC.
+
+			// +1 for this HTLC.
+			let local_fee_cost_msat = self.next_local_commit_tx_fee_msat(Some(1));
+			if self.value_to_self_msat < self.local_channel_reserve_satoshis * 1000 + local_fee_cost_msat {
+				return Err(ChannelError::Ignore("Cannot receive value that would put us under local channel reserve value"));
+			}
 		}
+
+		let pending_value_to_self_msat =
+			self.value_to_self_msat + htlc_inbound_value_msat - removed_outbound_total_msat;
+		let pending_remote_value_msat =
+			self.channel_value_satoshis * 1000 - pending_value_to_self_msat;
+		let channel_reserve_msat =
+			Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis) * 1000;
+		let remote_fee_cost_msat = if self.channel_outbound { 0 } else {
+			// +1 for this HTLC.
+			self.next_remote_commit_tx_fee_msat(Some(1))
+		};
+		// Note that rather than subtracting the HTLC amount and remote fee cost from the remote's
+		// pending value, which would be more intuitive to read,  we add it on the other side of
+		// the equation to prevent overflow. We need to be very careful about avoiding overflows
+		// here because they can lead to problems such as unnecessary channel closes.
+		if pending_remote_value_msat < channel_reserve_msat + msg.amount_msat + remote_fee_cost_msat {
+			return Err(ChannelError::Close("Remote HTLC add would put them under remote reserve value"));
+		}
```

Comment:
Not quite sure I followed earlier what would cause overflow (or underflow?). @valentinewallace or @TheBlueMatt Could you help me understand? Would it be from a bad `msg.amount_msat`?

Reply:
Not meaning to state the obvious, but I thought it'd just happen if `msg.amount_msat + remote_fee_cost > pending_remote_value` ? So it could happen from them trying to send too much, or the onchain fee being too high?

Reply:
Right.. I (mistakenly?) read @TheBlueMatt's comment to be about integer overflow/underflow. At least, usually re-ordering mathematical operations is used to prevent that. So that is what my question was regarding.

If the purpose is instead to _prevent_ the remote side from going over the channel reserve + fee cost (actually, _under_ it, right?), then I'm not quite sure I follow the argument regarding moving the variables to different sides of the equation. This is prevented no matter how we arrange the equation.

Reply:
I was thinking it was related to the integer overflow/underflow case. Basically trying to avoid the overflow in this situation: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fa10698fc7a75cfbc6b01b3d4b8144ee

Updated my code comment though because I think I was incorrect about the consequences of the overflow.

Reply:
Thanks for the explanation and example! I've refined my opinion on this now. :)

I see the need to be extra cautious here, but I think it can be done without loss of code clarity, which is beneficial in the long-run especially as the spec changes. Currently, the spec states:

> A receiving node:
> - receiving an `amount_msat` that the sending node cannot afford at the current `feerate_per_kw` (while maintaining its channel reserve):
>     - SHOULD fail the channel.

I don't think the wording is clear because it's combining three separate checks:
1. Does the sender have enough funds to make the payment?
1. Will the sender have enough funds to pay for fees (if applicable)?
1. Will the sender maintain their channel reserve?

The need to be cautious about overflow arises because we are performing all three checks at once, just as the spec is worded. However, they could be separated as:

```diff
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 2d76dd3..4a145dd 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1802,16 +1802,21 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
                        self.value_to_self_msat + htlc_inbound_value_msat - removed_outbound_total_msat;
                let pending_remote_value_msat =
                        self.channel_value_satoshis * 1000 - pending_value_to_self_msat;
-               let channel_reserve_msat =
-                       Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis) * 1000;
+               if pending_remote_value_msat < msg.amount_msat {
+                       return Err(ChannelError::Close("Remote HTLC add would overdraw remaining funds"));
+               }
+
                let remote_fee_cost_msat = if self.channel_outbound { 0 } else {
                        // +1 for this HTLC.
                        self.next_remote_commit_tx_fee_msat(1)
                };
-               // Note that rather than subtracting the HTLC amount and remote fee cost from the remote's
-               // pending value, which would be more intuitive to read,  we add it on the other side of
-               // the equation to prevent integer overflow.
-               if pending_remote_value_msat < channel_reserve_msat + msg.amount_msat + remote_fee_cost_msat {
+               if pending_remote_value_msat - msg.amount_msat < remote_fee_cost_msat {
+                       return Err(ChannelError::Close("Remote HTLC add would not leave enough to pay for fees"));
+               }
+
+               let channel_reserve_msat =
+                       Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis) * 1000;
+               if pending_remote_value_msat - msg.amount_msat - remote_fee_cost_msat < channel_reserve_msat {
                        return Err(ChannelError::Close("Remote HTLC add would put them under remote reserve value"));
                }
```

The advantage here is the checks are easy to understand as they incrementally build up and have clear explanations in the error messages. And no lengthy comment is needed. Asserts could be used between each check if even more caution is desired.

Ultimately, a channel balance could use its own abstraction to encapsulate this behavior. Not quite sure how to get there yet, but I may do a follow up PR to refactor some of `Channel` in this manner.

Reply:
That's way more readable. Updated